### ISSUE TITLE
Upgrade Zola Ember & Set Node Version

### ISSUE THREAD

**Issue by allthesignals** [NYCPlanning/labs-zola#1058]
# This issue is closed when Ember is version 3.19 and the app builds in Netlify for Node 14/16

# Development issues

## "yarn add node-sass never finish"

See workaround: https://github.com/yarnpkg/yarn/issues/5129#issuecomment-463726676

```bash
npm i node-sass
yarn add node-sass
yarn install
```

TODO: should switch out labs-ui, rework the sass-based styling dependency, and consider postcss to compile sass. Or, consider a static build of labs-ui styles... then do some overrides where necessary...

## Cannot find module '@babel/plugin-proposal-private-property-in-object'

Simply `yarn add` this dependency

### Build Error (fontawesome-svg-core): acorn.Parser.extend is not a function

Currently unknown....

## General Direction so far

A huge issue and barrier here is node-sass and foundation. Foundation-Sites has some known SemVer "violations" which make upgrading difficult and maintenance unsustainable.

Further, node-sass is a C dependency, which is high overhead for setting u pthe project.

I dealt with this in ZAP Search frontend by pulling in the _compiled_ "Labs UI" stylesheets as a single asset and building on top of that. Pre-processing was moved away from node-sass (deprecated) and towards PostCSS.

Because node-sass is deprecated (I believe it's now "Dart Sass" or something), we should move away from this and towards a high-level compiler (node-based). This will elminate the need for a C compiler for _SASS_ (which is wild). I think for some orgs with extremely complicated stylesheets, a fast compiler is very nice, but that is not our situation currently.

What this means is needing to move some things away from the labs-ui shared addon. I dealt with this in ZAP Search Front End by pulling in all the shared components in the consumer repo. This creates duplication, yes, but reduces overheard. I've heard this advice from other organizations dealing with addon-related upgrade woes.

I'm open to suggestions on this. The other option is to "fix" the labs-ui dependency, but I'm concerned with the unnecessary overhead this may create. I'll spike this path to see if it's a possible "fix-all" for consuming repos!

**Comment by TylerMatteo** [comment]
I totally hear you on upgrading labs-ui being a headache but I really do not want us pulling stuff out of shared repositories. The Labs portfolio doesn't do the best job of "sharing code" (be it through npm packages, RESTful apis, or other means) as it stands today, so I'd rather not add to that by pulling out and duplicating _the shared code we do have_. You're on the right track with the sass stuff but maybe I can clear things up - in this context, "SASS" refers to a language-agnostic "spec" that has been implemented in several languages. As you pointed out, node-sass is a thin wrapper around the C implementation. "Dart Sass" is an implementation written in Dart that has been the "default" one for a little while now - so it gets new features the soonest and generally has the best support. It's probably worth looking into if we can move to [this package](https://www.npmjs.com/package/sass) which is the JS distribution of Dart Sass (as opposed to the one we currently use which comes from another npm package named `node-sass`)

**Comment by TylerMatteo** [comment]
Fwiw, it looks like our node-sass dependency comes from `broccoli-sass-source-maps` which, in turn, comes from `ember-cli-sass` version 7.2.0. That's a whopping 4 major version behind the current version of `ember-cli-sass` but the most recent version of that dep [depends on](https://github.com/adopted-ember-addons/ember-cli-sass/blob/master/package.json#L64) `sass`, the dart-sass-based package.

**Comment by allthesignals** [comment]
> You're on the right track with the sass stuff but maybe I can clear things up - in this context, "SASS" refers to a language-agnostic "spec" that has been implemented in several languages. As you pointed out, node-sass is a thin wrapper around the C implementation.

Yes this is my understanding.

> The Labs portfolio doesn't do the best job of "sharing code" (be it through npm packages, RESTful apis, or other means) as it stands today, so I'd rather not add to that by pulling out and duplicating _the shared code we do have_.

That's what I'm trying first.

> It's probably worth looking into if we can move to [this package](https://www.npmjs.com/package/sass) which is the JS distribution of Dart Sass (as opposed to the one we currently use which comes from another npm package named `node-sass`)

I'll try that first.

> That's a whopping 4 major version behind the current version of `ember-cli-sass` but the most recent version of that dep [depends on](https://github.com/adopted-ember-addons/ember-cli-sass/blob/master/package.json#L64) `sass`, the dart-sass-based package.

That's promising.

**Comment by TylerMatteo** [comment]
Yeah you're correct about the upgrade being the main objective, I just want us to at least make a serious effort to do so without introducing other forms of tech debt into the portfolio. That being said, I'm not necessarily expecting you to reintegrate dependencies where they've already been split out, such as ZAP search - my hope is that would be relatively easy for the team to handle once all the upgrades are complete.

**Comment by allthesignals** [comment]
Totally separate, but _wow_, layers-api:

<img width="818" alt="image" src="https://user-images.githubusercontent.com/5004319/176201462-25c177a1-6aaf-47ed-8bd7-2bcf75cb0931.png">

I don't even know if that application is set up to support compression...

**Comment by allthesignals** [comment]
Okay, pulling down labs-ui to develop locally and hopefully solving some of this stuff

**Comment by allthesignals** [comment]
Promising: swapping in the new dart-baesd ember-cli-sass drops node-gyp and node-sass as expected:

<img width="473" alt="image" src="https://user-images.githubusercontent.com/5004319/176203328-ffc9d34b-4f4d-4aaa-a622-d11ec596769c.png">

I've opened a PR for the upgrade of labs-ui, but I'm going to test it locally as well to see how a version bump of it might affect consumer apps: https://github.com/NYCPlanning/labs-ui/pull/78.

~Having a little trouble testing this locally with `yarn link`... I recall symlinking doesn't actually fully simulate the way these packages work... ~ ~edit: [oh](https://github.com/yarnpkg/yarn/issues/6880#issuecomment-981581604)~ still not able to test consuming app locally... I vaguely recall a tool that did this... [here it is](https://github.com/wclr/yalc)... this works a little better when debugging dependency/version issues in packages. Okay, that's revealing that I need to run dependency-lint on labs-zola to figure out why it's resolving for two different versions of `broccoli-sass-source-maps`. More to come...

**Comment by allthesignals** [comment]
Stepping back and running dependency lint here's where we are:

<img width="654" alt="image" src="https://user-images.githubusercontent.com/5004319/176243861-a20d684b-3cde-4129-8a10-c62bf5826c4e.png">

I have a branch for another missing dependency that might've popped up from some dependency drift or node versioning (who can say): [mg/escape-dependency-hell](https://github.com/NYCPlanning/labs-zola/tree/mg/escape-dependency-hell)

Fixed dependency lint issue by upgrading fort-awesome...

Am I being trolled?
<img width="569" alt="image" src="https://user-images.githubusercontent.com/5004319/176245333-a4b2d6e0-4028-49f9-b62e-ec5bc1840a11.png">

It's crazy, it almost seems to resolve stuff non-deterministically (regenerating yarn.lock for example). No matter what it seems to resolve @fortawesome/ember-fontawesome with different versions. Fun! Going to punt on that path for now...

Okay one source of conflicting versions: labs-ui and labs-ember-search both specific ember-power-select as dependencies. Depending on the version, labs-ui asks for power-select 4. ember-search is still on version 2. Hence, upgrades lead to build time errors... turning towards upgrading ember-search now...

**Comment by allthesignals** [comment]
All sorts of strange behavior on labs-ember-search, and that one has a small number of dependencies. I'm going to keep chipping away that one and see what I learn... 

The  (ember-cli-search) package's current ember-cli version requires node 10. Upgrading ember-cli griefs me so far... 

Made some progress:

1. On Node 12
2. On ember-cli 3.10 (no more Active LTS warning)
3. Moved deps that didnt' need to be in "dependencies" into "devDevependencies"

In order to upgrade ember-power-select, I will need to upgrade ember-source due to the [ember-basic-dropdown](https://github.com/cibernox/ember-basic-dropdown) dependency...

**Comment by TylerMatteo** [comment]
Thanks for documenting all these trials and tribulations 😆 The dependency path to ember-basic-dropdown definitely rings some bells to the last time I tried to untangle this. Sounds like you're making some progress?

**Comment by TylerMatteo** [comment]
As you're digging into labs-ember-search, just want to point you at [this PR](https://github.com/NYCPlanning/labs-ember-search/pull/35) I did back in December. I had to revert some upgrades so that I can release a version that was still compatible with older apps but included some data-related updates we needed. Not sure if it's relevant to your work here, but I'd imagine you might end up re-implementing some of the changes I reverted in this PR so just wanted to make sure you're aware.

**Comment by allthesignals** [comment]
@TylerMatteo okay, thank you for that context! I'm trying to avoid upgrading things too far. My goal is to test the two main dependencies in consuming apps before committing to them. I've upgraded to 3.11.

All that to say... I can see how all this dependency mess has kept everyone and everything "stuck" in a really bad way! Going to keep digging...

**Comment by allthesignals** [comment]
Let me know if you prefer I not document things here, just need to keep track of where I left things in case I need to switch context to something else.

Currently: using yalc to test labs-ember-search with other consuming apps and noting common pitfalls...

**Comment by allthesignals** [comment]
Going from labs-ui 0.0.25 to 1+ is going to be pretty bad because labs-zola actually directly extends some of the components in there.

Two paths:
1. I'm going to see if I can fix the way it extends those components.
2. fork from labs-ui 0.0.25

**Comment by allthesignals** [comment]
Path 1 leads to some blockers that are usually solved by codemods, however, there is some behavior going on in the linked code below that I think will probably totally block a full upgrade to Ember 3.19:

https://github.com/NYCPlanning/labs-zola/blob/develop/app/templates/components/labs-ui/layer-groups-container.hbs#L1-L22

I'll see where I can get with this. I at least have a solution path towards 3.11.

**Comment by TylerMatteo** [comment]
>Let me know if you prefer I not document things here, just need to keep track of where I left things in case I need to switch context to something else.

I think documenting all this stream-of-consciousness stuff here is great. Once you have any cleaned up PRs that are ready to review, you can describe what you actually ended up doing in those (within reason, don't need a book for each PR).

I'm totally onboard with you doing multi-step upgrades, such as going to 3.11 before continuing on to 3.19 if you think that will be helpful. I'd imagine it might help break up the work. The extending thing in Zola is a little concerning. Do you think that is an artifact of a previous attempt to move some of that labs-ui code out of the library and duplicate it into Zola directly?

**Comment by allthesignals** [comment]
> I think documenting all this stream-of-consciousness stuff here is great. Once you have any cleaned up PRs that are ready to review, you can describe what you actually ended up doing in those (within reason, don't need a book for each PR).

Great!

> The extending thing in Zola is a little concerning. Do you think that is an artifact of a previous attempt to move some of that labs-ui code out of the library and duplicate it into Zola directly?

I think it's just an artifact of bad design... I _think_ I wanted to use some [composability](https://github.com/miguelcobain/ember-composability-tools) with the existing labs-ui components (I wanted programmatic access to rendered children of "layer-group-containers", for example, "Zoning and Land Use":
<img width="288" alt="image" src="https://user-images.githubusercontent.com/5004319/176472581-15775591-73e0-43ca-bf81-cb9e04b9abf6.png">

 but I think the problems are fixable. I'm going to try to keep things away from "refactor" zone and firmly inside of "codemod/syntax updates".

**Comment by allthesignals** [comment]
Note to self: assess ZoLa deprecation warnings _if_ we get to 3.19.

**Comment by TylerMatteo** [comment]
When you feel like you have your thoughts straight (which I know is hard to do with this stuff), can you give me a brief summary of where you're at with all of this? In particular, I'm curious which labs-built dependency you think need upgraded before we can upgrade Zola (and, presumably, the other front ends). It sounds like at least labs-ember-search, labs-ui, and ember-mapbox-composer would be on that list? Any others?

**Comment by allthesignals** [comment]
I'm still in the midst of discovery mode so I won't have a concise answer for you on that just yet but give me some more time and I'll see where I've landed. There are a lot of moving parts (this is less straightforward because of the requirement to keep labs-built addons/dependencies rather than pulling them into consumer apps).

**Comment by TylerMatteo** [comment]
> I tried that, the problem is that things were switched to pure glimmer components, which is a breaking change (hence the major semver). And the handlebars syntax for some of the extended components do not work with pure glimmer components. 

Gotcha. Is it correct to say that labs-ui could be upgraded to ember 3.19 _without_ switching to pure glimmer components? In other words, did Andy switch to glimmer components _so that he could upgrade_ or was it just taking advantage of a new feature that the upgrade enabled? 

> I am going to pull in (duplicate) the addons temporarily while upgrading ZoLa.

Seems reasonable enough to me. Maybe we can squash commits with the duplicated components before we go to master, just to keep the commit history less confusing.

**Comment by allthesignals** [comment]
> Gotcha. Is it correct to say that labs-ui could be upgraded to ember 3.19 _without_ switching to pure glimmer components? In other words, did Andy switch to glimmer components _so that he could upgrade_ or was it just taking advantage of a new feature that the upgrade enabled?

The latter — taking advantage of new features. Within 3+, "classic" components are still backwards compatible so we can upgrade without switching to glimmer components.




> Seems reasonable enough to me. Maybe we can squash commits with the duplicated components before we go to master, just to keep the commit history less confusing.

Sure, I'm SO FAR getting into a rhythm with the "ejected" labs-ui... upgrade 1 minor version, run tests, pass, repeat. (The only "test" that fails is DependencyLint). Knock on wood!

So far, I've _only_ ejected labs-ui... this _may_ help me uncover what the source of pain is so that I can go back and redo my commit history _without_ the ejection. We shall see. This wil inform upgrades for the other apps.

**Comment by allthesignals** [comment]
What this upgrade process will do is it will start throwing a bunch of new deprecation warnings in the development console. The next step would be to address those to prepare the app to be upgraded to 4.

**Comment by allthesignals** [comment]
This is going well. Currently on 3.13.

**Comment by allthesignals** [comment]
https://github.com/NYCPlanning/labs-zola/pull/1059 

Ember 3.12 upgrade available on this branch.

However, I'm in a bit of a catch-22: tests fail due to dependency-lint issues. This is somehow related to the way versions are specified on labs-ui and labs-ember-search:
https://github.com/NYCPlanning/labs-ember-search/blob/master/package.json#L43
https://github.com/NYCPlanning/labs-ui/blob/27f75acec891f073067cbb907b7071596b7517d1/package.json#L22

I have tried to fix those issues, but they would seem to require upstream changes to labs-ui and labs-ember-search.

As a temporary stopgap, I _could_ configure dependency lint to [allow multiple versions](https://github.com/salsify/ember-cli-dependency-lint#allowed-versions).

## Notes

```
Unexpected token '.'


Stack Trace and Error Report: /var/folders/g0/l926clqx2zn2wgvvx_r8xs700000gp/T/error.dump.7e122204d4c2a78fe14fed00fc1e9d49.log
cleaning up...
```

Super unhelpful error! Stems from `ember-cli-fastboot-testing`.

**Comment by allthesignals** [comment]
[Landed on 3.13](https://github.com/NYCPlanning/labs-zola/tree/mg/escape-dependency-hell-3.13)! Automated tests fail for some reason, will fix that soon, but got a quick visual check on things. So, some real skeletons in the closet here... 

1. We have an addon, ember-mapbox-composer, that itself overrides some private methods in ember-mapbox-gl. That's bad. But even worse, labs-zola extends ember-mapbox-composer, which also overrides private stuff. So, things were failing because of private internals changed, overrides no longer worked... and I had to add some code to get these to trigger manually.
2. To get around the foundation-sites issue, I had to downgrade and pin the version of foundation-sites due to run-ins with this [deprecation warning/error combination](https://github.com/foundation/foundation-sites/issues/12053) issue. The same thing is done in labs-streets
3. ember-needs-async is seemingly abandoned and inserts specific versions of ember-data, causing yarn to resolve with multiple of ember-data. This causes weird inexplicable polyfills to fail/enter the picture. I don't know. Key was to get it to resolve to only one version of ember-data, and that meant forking ember-needs-async and making changes there.

**Comment by TylerMatteo** [comment]
Hey, I just submitted a review on your PR, sorry for the delay. I'm psyched to see it actually goes all the way to 3.19! I'm afraid I probably won't be much help when it comes to debugging tests but I'm happy to look into it next week.

**Comment by TylerMatteo** [comment]
Hey Matt, sorry I've been MIA. How's it going with trouble shooting those tests? I saw your email and am happy to schedule something for early next week if you still want to chat.

**Comment by TylerMatteo** [comment]
@allthesignals Just took another look at your PR and resolved a bunch of conversations. This is looking pretty close to being ready for a merge into develop to me. The one thing I'd like for us to do is have those commit hash dependencies be pointing to commits in repos owned under the NYCPlanning Github org. I'm out of the office next week but when I'm back we can try to get your ember-mapbox-composer PR (please take a look at [my comment](https://github.com/NYCPlanning/ember-mapbox-composer/pull/35#issuecomment-1212456232) on that PR) merged in and publish a version so that we don't need the hash at all. As for ember-needs-async, I can set up a fork of under our GH org when I get back. Almost there! Hopefully you've been able to make some progress on applicant portal backend and aren't totally blocked by your Zola PR languishing a little longer.

**Comment by TylerMatteo** [comment]
Ok I forked [ember-needs-async](https://github.com/NYCPlanning/ember-needs-async) and gave you access so you should be able to push your branch directly if you'd rather do that than PRing from your own fork, up to you. Once we have that merged, you can update the hash here to the one under the NYCPlanning fork

I left a comment on your [composer PR](https://github.com/NYCPlanning/ember-mapbox-composer/pull/35#issuecomment-1225945810) as well. Just need to tweak `engines` and `npm version` that PR. Once that's merged, I'll publish the new 0.2.5 version to npm so you hopefully won't need the hash for that one.

**Comment by TylerMatteo** [comment]
@allthesignals Please see my previous comment. Once you update your composer PR with those changes and open a PR against our new ember-needs-async fork, I can see about getting these changes out the door.

**Comment by TylerMatteo** [comment]
Hey @allthesignals, how difficult do you think it would be to remove the dependency on `ember-needs-async`? It looks like it's only used in `carto-data-provider` and I would love to refactor out an abandoned dependency if we can. Could someone refactor that to just use ember-concurrency directly?

**Comment by TylerMatteo** [comment]
@allthesignals I see that you've updated the mapbox-composer dependency to 0.3.0 and updated the hash for needs-async, so those issues look good to me. If you would have to copy over all of that code to remove ember-needs-async, we may as well just keep using that fork. I was hoping there might be a way to rebuild that one component where we use it to have a one-off usage of ember-concurrency directly, but I don't want to keep moving the goal posts here.

A couple things:
* Have you recently rebased your branch onto develop? If not, now would be a good time to do so and make sure things still work and the tests still work locally
* When you say the tests work locally, does that include the acceptance, unit, and integration or just a subset of them? I could probably live with not having them all working for now, just want to know.
* Do these changes still remove lint-staged and husky? If so, could we revisit that before we merge? We have some Zola work planned for the current sprint and disabling those is going to make things messy on our end.

cc PR#[1059](https://github.com/NYCPlanning/labs-zola/pull/1059)

**Comment by TylerMatteo** [comment]
Node 14/Ember 3.19 upgrade is live  🥳
